### PR TITLE
Rewrite ExprName

### PR DESCRIPTION
**Pull request by APickledWalrus**
### Description
This PR rewrites ExprName. While the solution to the related issues (probably) *isn't ideal*, I would say that in this case, it's better to have something than nothing. I found that overall performance increased, especially in the first few iterations (although I don't know how accurate an in-game benchmark is haha).

---
**Target Minecraft Versions:**     Any
**Requirements:**     None
**Related Issues:**     #3103 #2069 (kind of)

### CODE REVIEW

File: src/main/java/ch/njol/skript/expressions/ExprName.java [hunk 1]
```diff
@@ -97,336 +96,151 @@
 	static final MethodHandle TITLE_METHOD;
 
 	static {
+		register(ExprName.class, String.class, "(1¦name[s]|2¦(display|nick|chat|custom)[ ]name[s])", "players/entities/itemtypes/inventories/slots");
```
Comment (FranKusmiruk) [inline src/main/java/ch/njol/skript/expressions/ExprName.java:99]: Slots are convertible to itemtypes, you may test whether removing slots from the accepted types causes any issues (changers primarily).

File: src/main/java/ch/njol/skript/expressions/ExprName.java [hunk 2]
```diff
@@ -97,336 +96,151 @@
 	static final MethodHandle TITLE_METHOD;
 
 	static {
+		register(ExprName.class, String.class, "(1¦name[s]|2¦(display|nick|chat|custom)[ ]name[s])", "players/entities/itemtypes/inventories/slots");
+		register(ExprName.class, String.class, "(3¦(player|tab)[ ]list name[s])", "players");
```
Comment (FranKusmiruk) [inline src/main/java/ch/njol/skript/expressions/ExprName.java:100]: They are different patterns. I'd argue this is a misuse of parser marks considering you should be doing a pattern check here but parser marks are in some way a tool to avoid pattern-checking code as well. What do the rest think about this? Should the "misuse" if I can even call that be overlooked for sake of code cleanliness?

File: src/main/java/ch/njol/skript/expressions/ExprName.java [hunk 3]
```diff
@@ -97,336 +96,151 @@
...
-
-	static {
-		for (NameType n : NameType.values())
-			register(ExprName.class, String.class, n.pattern, n.getFrom());
-	}
-
-	@SuppressWarnings("null")
-	private NameType type;
+	/**
+	 * 1 = "name",
```
Comment (FranKusmiruk) [inline src/main/java/ch/njol/skript/expressions/ExprName.java:112]: This implementation detail might be better as a normal multi-line comment rather than a javadoc.

File: src/main/java/ch/njol/skript/expressions/ExprName.java [hunk 4]
```diff
@@ -97,336 +96,151 @@
...
-			if (Slot.class.isAssignableFrom(getExpr().getReturnType())) {
-				for (Slot s : (Slot[]) getExpr().getArray(e)) {
-					ItemStack i = s.getItem();
-					type.set(i, name);
-					s.setItem(i);
+		String name = delta != null ? (String) delta[0] : null;
+		for (Object o : getExpr().getArray(e)) {
+			if (o instanceof Player) {
+				switch (mark) {
+					case 2: ((Player) o).setDisplayName(name != null ? name + ChatColor.RESET : ((Player) o).getName());
```
Comment (FranKusmiruk) [inline src/main/java/ch/njol/skript/expressions/ExprName.java:182]: missing break statement (even if the following code is unreachable, the compiler throws a warning for it so)

File: src/main/java/ch/njol/skript/expressions/ExprName.java [hunk 5]
```diff
@@ -97,336 +96,151 @@
...
+				if (mark == 2 || mode == ChangeMode.RESET) // Using "display name"
+					((Entity) o).setCustomNameVisible(name != null);
+				if (o instanceof LivingEntity)
+					((LivingEntity) o).setRemoveWhenFarAway(name == null);
+			} else if (o instanceof ItemType) {
+				ItemType i = (ItemType) o;
+				ItemMeta m = i.getItemMeta();
+				m.setDisplayName(name);
+				i.setItemMeta(m);
+			} else if (o instanceof Inventory ) {
```
Comment (FranKusmiruk) [inline src/main/java/ch/njol/skript/expressions/ExprName.java:196]: space before ending parenthesis

File: src/main/java/ch/njol/skript/expressions/ExprName.java [hunk 6]
```diff
@@ -97,336 +96,151 @@
...
+				Inventory inv = (Inventory) o;
+
+				if (inv.getViewers().isEmpty())
 					return;
-				}
+				// Create a clone to avoid a ConcurrentModificationException
+				List<HumanEntity> viewers = new ArrayList<>(inv.getViewers());
```
Comment (FranKusmiruk) [inline src/main/java/ch/njol/skript/expressions/ExprName.java:202]: Don't create a clone, use its iterator instead. Or are you even modifying the list to generate a CME? I don't see any code that could do it.

File: src/main/java/ch/njol/skript/expressions/ExprName.java [hunk 7]
```diff
@@ -97,336 +96,151 @@
...
+					is.setItemMeta(m);
+					s.setItem(is);
+				}
 			}
 		}
 	}
 
 	@Override
-	public Class<String> getReturnType() {
+	public Class<? extends String> getReturnType() {
```
Comment (FranKusmiruk) [inline src/main/java/ch/njol/skript/expressions/ExprName.java:232]: Why did you change this?

File: src/main/java/ch/njol/skript/expressions/ExprName.java (hunk 6 above)
Comment (APickledWalrus) [reply src/main/java/ch/njol/skript/expressions/ExprName.java:202]: A CME can be encountered when opening the inventory later on, at least that’s what I encountered.

File: src/main/java/ch/njol/skript/expressions/ExprName.java (hunk 7 above)
Comment (APickledWalrus) [reply src/main/java/ch/njol/skript/expressions/ExprName.java:232]: I kind of worked from the ground up, so this was unintentional.

File: src/main/java/ch/njol/skript/expressions/ExprName.java (hunk 2 above)
Comment (APickledWalrus) [reply src/main/java/ch/njol/skript/expressions/ExprName.java:100]: I wasn’t sure how a pattern check here would work since I’m calling register twice. Would the tablist pattern return "2" for matched patterns? I’m not opposed to doing a pattern check at all, I’m just not should if/how it would work since I’m calling register twice.

File: src/main/java/ch/njol/skript/expressions/ExprName.java (hunk 6 above)
Comment (FranKusmiruk) [reply src/main/java/ch/njol/skript/expressions/ExprName.java:202]: Ah, yeah, this is alright then.

File: src/main/java/ch/njol/skript/expressions/ExprName.java (hunk 2 above)
Comment (FranKusmiruk) [reply src/main/java/ch/njol/skript/expressions/ExprName.java:100]: Talked about it in Discord; since Skript doesn't treat the second registration as an additional pattern but rather, a new expression info and this being unhandled behaviour, it's better to leave it as is for now.

File: src/main/java/ch/njol/skript/expressions/ExprName.java (hunk 1 above)
Comment (APickledWalrus) [reply src/main/java/ch/njol/skript/expressions/ExprName.java:99]: Tested this. Works with the getter but not the changer, presumably because you have to update the slot's item(?).
